### ISSUE TITLE
[HOLD for payment 2022-03-02] [$250] Add the ability to configure the distance tracking `rate` and `unit` on the Free Plan

### ISSUE THREAD

**Issue by trjExpensify** [Expensify/App#6213]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Deliverables
1. Update the `Reimburse receipts` workspace page to be titled `Reimburse expenses` instead

![image](https://user-images.githubusercontent.com/16232057/137988250-2b7e3068-7e1a-4d80-bb1a-6f7f588a1002.png)

2. Add a new sub-section to the `reimburse expenses` page with the header `Track distance` and the copy: `Set the per mile/km rate and choose a default unit to track.`
3. Add the `Rate` field to allow for the rate to be changed by calling the command `Policy_CustomUnit_Update`
4. Add the `Unit` dropdown to be able to track distance in `Miles` (default) or `Kilometers` and allow for it to be changed by calling the command `Policy_CustomUnitRate_Update`

![image](https://user-images.githubusercontent.com/16232057/140099384-2502cf86-47be-452f-b088-ce240c65004d.png)

## Platform:
New feature, to be available on all platforms. 

**Notes/Photos/Videos:** Added in-line above. Here is the SVG asset for the car icon: [GPS Track.svg.zip](https://github.com/Expensify/Expensify/files/7413560/GPS.Track.svg.zip)
**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/181768
**Issue reported by:** trjexpensify
**Slack conversation:** https://expensify.slack.com/archives/C03U7DCU4/p1634665726096500
**Upwork URL:** https://www.upwork.com/jobs/~01143e8c4bd3284e03
[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @CortneyOfstad (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by sonialiap** [comment]
moving user from the other GH issue
harvey@connectdaycare.co.uk
https://www.expensify.com/concierge/#/chat/6457965

**Comment by deetergp** [comment]
I guess I'll stick around for the proposals then, since I'm already here.

**Comment by marktoman** [comment]
### Proposal

I get 404 on the car icon, so a rocket will have to do for now :)

#### Desktop
https://user-images.githubusercontent.com/47689923/140427318-e4dce09d-06d0-4752-83d7-9c0d386dee88.mp4

#### Mobile
https://user-images.githubusercontent.com/47689923/140427928-c4b57159-e83e-4679-aa4f-2568665fe546.mp4

**Comment by marktoman** [comment]
@parasharrajat ?

**Comment by parasharrajat** [comment]
@marktoman Thanks for taking interest in this job. But the videos you added earlier are not useful to us at all. They did not tell us how you are going to achieve this. 
Simply putting a solution video is never considered a good proposal. Also, showing diff does not tell everything we always suggest proposing an approach here not the complete solution. Could you please add an explanation of how you are going to achieve this? You can also look at old proposals from other contributors to get an idea. 

Here are a few things I prefer in a proposal.
1. Not too much detail.
2. Straight steps.
3. References to existing code that is going to be affected or causing the issue, etc.
4. Optional (a video) , helps in getting a visual understanding of some type of issues.

**Comment by marktoman** [comment]
2. Straight steps:
- Add functions to `API.js` that will call the requested commands over the network.
- Add functions, likely to `User.js` that will call those APIs, handle errors and potentially merge with Onyx.
- Add two components to `WorkspaceReimburseNoVBAView` for `Unit` and `Rate` representation according to the image.
- Bind the change methods to call the update functions we added earlier to send changed values over the network
- Add the requested strings in `en.js` with the requested keys. Use `translate` when reading them.
- Style the two view components according to the requested image in `styles.js`.
- Add a new icon to `components/Icon/Illustrations`

3. No issues expected given it is an addition (well seen in the diffs).

**Comment by trjExpensify** [comment]
I'll defer to @deetergp for reviewing the proposed solution. In the meantime Mark, can you apply to the job on Upwork as well? Thanks!

**Comment by marktoman** [comment]
@trjExpensify @deetergp Applied.

**Comment by mallenexpensify** [comment]
We are placing all non-critical issues on hold while we're on a company offsite.  The hold is expected to be lifted on 11/19 (but could go longer).  For any PRs that are submitted during the hold, we will add a $250 bonus payment.

**Comment by deetergp** [comment]
In general, @marktoman's proposal looks good. The one thing I'd say needs to be changed is the second bullet, as those update methods should probably be in `Policy.js` rather than `User.js` since they are making `Policy_*`  API calls on the back end. (Workspaces are essentially Policies)

**Comment by marktoman** [comment]
@deetergp Thank you and understood. Should I proceed with implementation?

**Comment by deetergp** [comment]
@marktoman Let's do it!

**Comment by trjExpensify** [comment]
Perfect. Hired on Upwork! 👍

**Comment by marktoman** [comment]
@trjExpensify Accepted, thank you!

**Comment by marktoman** [comment]
I am still getting 404 for the SVG file: https://github.com/Expensify/Expensify/files/7413560/GPS.Track.svg.zip

Should I create a PR without the icon in a few days, or do you think it will be available by then?

**Comment by trjExpensify** [comment]
@michelle-thompson @shawnborton any ideas?

**Comment by shawnborton** [comment]
Try downloading here: [GPS.Track.svg.zip](https://github.com/Expensify/App/files/7546990/GPS.Track.svg.zip)

**Comment by marktoman** [comment]
@shawnborton That works, thanks!

**Comment by marktoman** [comment]
I struggle a bit to understand how this fits into the existing policy concept. When I call `loadFullPolicy` for the `policyID` passed to `WorkspaceReimburseNoVBAView.js`, it already contains two arrays: `customUnits` and `rates`. Are the `Policy_CustomUnit_Update` and `Policy_CustomUnitRate_Update` commands related to these values? If so, should I just use the values from the first or last elements for the new text fields?

![customUnits](https://user-images.githubusercontent.com/47689923/142957978-9ea9ccd9-cf6f-4b85-885a-f05a110e9748.png)

Another thing is that the icon is created for the 100x100px view, but the `WorkspaceSection` uses 80x80px. Changing width and height on the `svg` element pushes the icon to the bottom right corner because each `path` seems to have an absolute value.

![image](https://user-images.githubusercontent.com/47689923/142958776-643e619e-068d-40e4-9602-e287e12bd227.png)

**Comment by shawnborton** [comment]
For the svg icon, maybe we need to manually scale down to 80x80 and then re-export from there? cc @michelle-thompson - I took a look at the [other product illustrations](https://github.com/Expensify/App/blob/main/assets/images/product-illustrations/bank-user--green.svg?short_path=676d05c#L3-L4) and looks like they were exported at 80x80

**Comment by michelle-thompson** [comment]
Lemme know if this works!
[gps-tracking.svg.zip](https://github.com/Expensify/App/files/7589525/gps-tracking.svg.zip)

**Comment by marktoman** [comment]
@michelle-thompson Perfect.

**Comment by marktoman** [comment]
Should I ignore these values? I seem to be missing some info.

**Comment by deetergp** [comment]
> Should I ignore these values? I seem to be missing some info.

Do you mean, should you ignore the default values passed with the policy? If that is what you are asking, then the answer is no.

**Comment by marktoman** [comment]
I was referring to [my previous question](https://github.com/Expensify/App/issues/6213#issuecomment-976093214). There is an `customUnits` array of where each element has an array of rates. Unless I am missing something, there is nowhere specified which element should be associated with the single `Rate` text field I have added.

**Comment by deetergp** [comment]
Ahhhh, I understand now, thanks for clarifying. 👍  

Yes, just take whichever unit is at index `0` for that array as they will only ever have the one rate available to them on that plan.

**Comment by marktoman** [comment]
I see, thanks!

**Comment by deetergp** [comment]
Heya @marktoman, where are we with this one?

**Comment by marktoman** [comment]
Hi @deetergp, if this is a priority, I will do this over this weekend. I think I have all the info.

**Comment by trjExpensify** [comment]
Hey Mark! This is definitely one of the most requested enhancements to the free plan at the moment, so it would be great if we could square this away please.

**Comment by marktoman** [comment]
@deetergp Where can I find the `Policy_CustomUnit_Update` API signature? Calling it with `policyID` and `unit` returns the following error:
```
Argument 3 passed to PolicyAPI::updateCustomUnit() must be of the type array, null given, called in /git/releases/expensify.com/cf6d815/api.php on line 1692
```

**Comment by deetergp** [comment]
@marktoman I am pretty sure it isn't documented anywhere, but that third parameter is the last modified date. Internally when we call `Policy_CustomUnit_Update` we usually send it `policy.lastModified || null`. Give that a shot and see if it works.

**Comment by marktoman** [comment]
@deetergp How should the call look like then? All policy calls pass an object to their API methods. I have tried specifying the `lastModified` property with various values, also multiple values for other properties that I do not know how should be named (just guesses based on the other calls) and always get the exact same error.

**Comment by marktoman** [comment]
Please provide the missing spec. The command name alone with no payload info is not very useful.

If this is a priority task, note that I cannot interrupt my other work anytime I please, hence these delays may compound.

**Comment by deetergp** [comment]
I see what's going on. You need to be passing a JSON object containing the unit data to the API. That JSON is converted to a PHP array before being sent to the method in the error message. I dropped a log line in that PHP api call and changed the the unit on one of our legacy policies and the individual parameters looked something like:
```
policyID: '004C88207FA3EA60' 
customUnit: '{"customUnitName":"Distance","customUnitID":"61ba4fb51e01e","attributes":{"unit":"km"}}' 
lastModified: '1639600234184791'
```

Hopefully that helps?

**Comment by marktoman** [comment]
@deetergp Thanks a lot, that works. Based on the unit logic, I was able to figure out the unit rate structure as well.

**Comment by marktoman** [comment]
@deetergp Please let me know if it does what is required. I will then fill out the QA steps etc.

**Comment by deetergp** [comment]
Hey @marktoman we generally require the testing/qa steps prior to doing the review. Would you please take a moment to fill them out?

**Comment by trjExpensify** [comment]
👋 Asked for an update on the PR directly, Melvin. It would be great to get this squared away this week, as it has been 2 months since we opened the issue.

**Comment by deetergp** [comment]
Mark updated the PR with testing instructions, but it did not pass testing. Waiting on fixes.

**Comment by marktoman** [comment]
Hi Scott, I have found the underlying issue (see my response on the PR), which is not necessarily related to the PR. Can you check it out?

**Comment by trjExpensify** [comment]
👋 Hi Melvin, looks like Mark is waiting for feedback from @deetergp on the PR [here](https://github.com/Expensify/App/pull/6839#issuecomment-1017880782) and we should be good to go. 👍

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.36-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6839


If no regressions arise, payment will be issued on 2022-02-14. :confetti_ball:

**Comment by trjExpensify** [comment]
👋 I just tested this and I think we have a problem, @marktoman & @deetergp. The default IRS rate for newly created policies should be `0.585`:

![image](https://user-images.githubusercontent.com/16232057/152856671-dd948963-23f2-4484-b870-49ec7e1a0db7.png)

Whereas in NewDot this appears as `58.50`:

![image](https://user-images.githubusercontent.com/16232057/152856934-5c499c49-45f2-4fc9-8a4f-45294710c8e5.png)

Also, when I change it to `0.585` we revert it back to two decimals in the UI (`0.58`):
![newdot_mileage](https://user-images.githubusercontent.com/16232057/152859150-6931ccf5-0ce2-4a8c-9e93-3fc9cb8287c0.gif)

Then when I create a mileage expense with the rate set at `0.58` in the workspace settings, it appears as though the rate used on the expense is `0.0059`:

![image](https://user-images.githubusercontent.com/16232057/152858343-be1de9be-9c76-4e33-9681-165b5e62aa30.png)

**Comment by trjExpensify** [comment]
Removing the `Awaiting payment` label until a fix is up for this, also moved the issue back to `Daily` as this is on production and going to cause confusion.

For the PR sent to fix it, I think we should add the `CP Staging` label. Curious for your thoughts on that too, Scott.

**Comment by marktoman** [comment]
Also, when I change it to 0.585 we revert it back to two decimals in the UI (0.58):
Should I change the correction to 3 decimal places?

The other errors do not seem to be related to this issue. After creating a new workspace, I receive the wrong number prior to touching it in any way:
![image](https://user-images.githubusercontent.com/47689923/152884508-d8218519-dbb8-497a-88b2-0f3c2afa2ef0.png)

**Comment by deetergp** [comment]
@marktoman I guess we need to allow for the possibility of three decimal places, but only show two when there isn't an expressly stated third decimal.

**Comment by trjExpensify** [comment]
>In other words if it's fifty-eight and a half cents, show it as 0.585, but if it's fifty cents, show it as 0.50 instead of 0.500.

👍 

>Should I change the correction to 3 decimal places?

On Expensify.com we allow for 4 decimal places, so I think we should do the same here as there will likely be a reason for that in the history of this feature (IIRC, some international territories set their mileage rate allowances at 4 decimals). 

>The other errors do not seem to be related to this issue. After creating a new workspace, I receive the wrong number prior to touching it in any way:

As for this, maybe @deetergp has an inclination as to what's up here? Just to clarify as well though, setting `58.50` as the `rate` in the workspace settings does use the correct `0.585` rate when actually creating the mileage expense: 

![image](https://user-images.githubusercontent.com/16232057/152975951-4c1a6269-7eed-469d-a3a0-ffdb535914ff.png)

![image](https://user-images.githubusercontent.com/16232057/152975888-cbdb8e8b-b3b5-4dc4-b958-1a34a33a3d5b.png)

**Comment by deetergp** [comment]
@trjExpensify Is that on production? In my local, if I set it `0.585` it comes back at `0.58` (Although it is actually `0.585` still, behind the scenes) which is a lot closer to being correct and a far cry from `58.50`.
![Screen Shot 2022-02-08 at 2 05 44 PM](https://user-images.githubusercontent.com/766197/153083841-28b5fa95-558a-4d51-9d8f-0a46aa9abe39.png)
![Screen Shot 2022-02-08 at 2 08 40 PM](https://user-images.githubusercontent.com/766197/153083960-6ae4f29a-378d-4b72-9b90-ea7ca6a55a3a.png)

**Comment by trjExpensify** [comment]
>@trjExpensify Is that on production?

Yep, Web on prod! Account if you want it: `tomrhysjones1+mileagetest@gmail.com`

> if I set it 0.585 it comes back at 0.58 (Although it is actually 0.585 still, behind the scenes) which is a lot closer to being correct and a far cry from 58.50.

Can you follow this through and create a mileage expense on a free plan policy once you've set the rate to `0.58` in the workspace settings. What result do you get? Pretty sure you're going to get this: 

>Then when I create a mileage expense with the rate set at `0.58` in the workspace settings, it appears as though the rate used on the expense is `0.0059`:

>![image](https://user-images.githubusercontent.com/16232057/152858343-be1de9be-9c76-4e33-9681-165b5e62aa30.png)

**Comment by deetergp** [comment]
Ahh, you're 100% correct @trjExpensify! I tested it locally and see the same results on the OldDot backend. I've addressed the issue [here](https://github.com/Expensify/App/pull/7639#issuecomment-1034374194) on the PR.

**Comment by trjExpensify** [comment]
Nice! 

>Does that make ~cents~ sense?

Top work 😂 👏

**Comment by iwiznia** [comment]
This added a few regressions. See [here](https://github.com/Expensify/App/pull/6839#discussion_r809230582).
@marktoman can you fix them?
@trjExpensify please hold on payment till the fixes are deployed.

**Comment by trjExpensify** [comment]
Hey all, back from OoO. Just to confirm, we're still waiting on [this PR](https://github.com/Expensify/App/pull/7639) to hit prod for the regression period to count (and the `Awaiting payment` label will be added?)

**Comment by trjExpensify** [comment]
Oh, @marktoman @deetergp. Did we lose the currency symbol (e.g `$`) along the way somewhere inside the `rate` field like here in the [OP mockup](https://github.com/Expensify/App/issues/6213#issue-1045037911) and the video from the proposal [here](https://github.com/Expensify/App/issues/6213#issuecomment-961475585)? 🤔

**Comment by marktoman** [comment]
@trjExpensify  It shows the currency when the field is blank. The text field component supports the main label (currently Rate) and the blank field label, but not a prefix label (meaning adding a prefix would interfere with the value itself).

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.39-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/7639


If no regressions arise, payment will be issued on 2022-03-02. :confetti_ball:

**Comment by trjExpensify** [comment]
>@trjExpensify It shows the currency when the field is blank. The text field component supports the main label (currently Rate) and the blank field label, but not a prefix label (meaning adding a prefix would interfere with the value itself).

Oh, hm. That's pretty hidden as there's always a default rate populated. @Julesssss @marcaaron @iwiznia, won't we need to be able to support a prefix label in the text field for N7?

**Comment by iwiznia** [comment]
I don't know what a prefix label is.

**Comment by iwiznia** [comment]
Oh, I see. I assume we will need it if we want to show the currency like that in an amount input.
Having said that, can we achieve this by adding the `$` character to the input's value and then not allowing the user to delete it? (thought that can still be weird for things like selecting the contents of the input)

**Comment by Julesssss** [comment]
Hmm yeah, that's a bit tricky to handle but I agree it will be necessary. As the BigNumberPad doesn't use a TextInput we have been able to use an alternate solution up to now (create a Unique Text Component for the currency symbol) -- but applying that solution to regular TextInputs isn't simple.

I wonder if we should create an external issue for creating a pre-fixable TextInput component, which allows users to copy **only** the numerical value. We need to decide first whether the symbol should be selectable or not via copy/paste, because that will drive the eventual solution.

**Comment by trjExpensify** [comment]
>I wonder if we should create an external issue for creating a pre-fixable TextInput component, which allows users to copy only the numerical value

I think this is a good idea. 

>We need to decide first whether the symbol should be selectable or not via copy/paste, because that will drive the eventual solution.

What are your thoughts here? I think it might be weird if you can select the entire value to copy/paste but can't backspace to also delete it. So with that, I'd probably lean towards not making it selectable?

**Comment by Julesssss** [comment]
> I think it might be weird if you can select the entire value to copy/paste but can't backspace to also delete it. So with that, I'd probably lean towards not making it selectable?

Yeah, I agree. Let's start with this. Will create an issue.

**Comment by trjExpensify** [comment]
Sounds good!

**Comment by trjExpensify** [comment]
Settled!
